### PR TITLE
Speed up shell integration

### PR DESCRIPTION
**Pull request by michaelkuhn**
This is likely to be a more controversial change; please see it more as a discussion starter.

The current shell integration script calls `spack-python` three times, which takes quite some time on our cluster (2-3s). Since the shell integration is loaded for every login, this can be quite annoying.

The change cuts down the amount of invocations to one for installations using the default dotkit and modules paths. On our cluster, this shortens the time needed to 0.7-0.8s.

### CODE REVIEW

**Comment by aprokop** [comment]
Without looking at the patch itself, I would certainly concur with the notion. Adding sourcing of the spack shell in `.bashrc` takes few seconds on opening a new terminal, which is intolerably slow for me. At the moment, I'm using a workaround where I only set `SPACK_ROOT` in my `.bashrc` while directly sourcing only as needed.

**Comment by citibeth** [comment]
You suggested this might be controversial.  Faster is definitely better; what's the downside of this change?

File: share/spack/setup-env.sh [hunk 1]
```diff
@@ -189,7 +189,23 @@ _sp_prefix=$(cd "$(dirname $(dirname $_sp_share_dir))" && pwd)
 _spack_pathadd PATH       "${_sp_prefix%/}/bin"
 
 _sp_sys_type=$(spack-python -c 'print(spack.architecture.sys_type())')
-_sp_dotkit_root=$(spack-python -c "print(spack.util.path.canonicalize_path(spack.config.get_config('config').get('module_roots', {}).get('dotkit')))")
-_sp_tcl_root=$(spack-python -c "print(spack.util.path.canonicalize_path(spack.config.get_config('config').get('module_roots', {}).get('tcl')))")
+
+_sp_use_custom_moduleroots="$(grep module_roots "${_sp_prefix}/etc/spack/config.yaml" > /dev/null 2>&1)"
```
Comment (adamjstewart) [inline share/spack/setup-env.sh]: There are other levels of configuration files you should check. A `config.yaml` could also be present in `~/.spack/`, `~/.spack/<platform>/`, `etc/spack/defaults/`, and `etc/spack/<platform>`. @citibeth plans on adding project level configuration files as well. @tgamblin am I missing any scopes?

**Comment by michaelkuhn** [comment]
@citibeth If the default paths should ever change, the shell integration would also need to be updated or the wrong paths might be used if the old dotkit and modules directories are still lying around. That is actually the only downside I can think of.

File: share/spack/setup-env.sh (hunk 1 above)
Comment (citibeth) [inline share/spack/setup-env.sh]: Project-level config scopes (on the command line) are not yet a supported feature.  If/when they are, I suppose we would have to specify them also on the command line to `setup-env.sh`?  That would be a bummer...

But I don't understand the issue very well at this point.

**Comment by tgamblin** [comment]
I actually don't think disagree with this.  Originally I attempted to make `setup-env.sh` NEVER call `spack location` and friends, but someone added these fixes and I didn't push back too hard. I would like to get to zero invocations again.

Observations:
1. I don't think reimplementing the config system in bash is ideal.
1. If you're still going to make one call to Python, would it make sense to add a special command for this, which could grab all the necessary info?  Then we could profile its startup and optimize.

Caveat: Python startup is probably the main bottleneck, particularly on shared filesystems.  I am not sure how much we can do to fix the baseline overhead if we still have one call to Python, but we could at least try.  If you want to go for zero python calls, we'd need to think a little harder.

**Comment by citibeth** [comment]
This might be helpful someday:

https://files.bemusement.org/talks/OSDC2008-FastPython/

**Comment by michaelkuhn** [comment]
Thanks for all the feedback!

I took a stab at improving this further. The new commit introduces a stripped down `spack-shell` command that output all the necessary information for the shell integration script. This still incurs quite some overhead but is actually faster than my previous approach by a factor of two.

**Comment by citibeth** [comment]
That sounds like a good approach to me.

**Comment by citibeth** [comment]
This looks like it might tie into the Spack Environments we are thinking
through.  For example... suppose I like to get my Spack environment by
doing a bunch of `module load` commands.  I have ways to generate a script
full of such commands.  Generating some shell integration as well in the
same script would tie in nicely.

**Comment by michaelkuhn** [comment]
@tgamblin I have also tried using a normal spack command (`spack shell`) but decided to go for this approach because it is much faster. It is especially noticeable on NFS (0.4s vs 0.8s on our cluster).

I have also tried the profiling option to find out where the additional time is taken but this seems to be very inaccurate. For a typical 0.8s run, `spack -p` shows something like this:
```
11963 function calls (11852 primitive calls) in 0.019 seconds
```

**Comment by alalazo** [comment]
@tgamblin @michaelkuhn ping

**Comment by michaelkuhn** [comment]
I just pushed an updated version that integrates the `spack-shell` logic into the `spack` binary and therefore eliminates all the redundancies. This still provides a speedup of 3, even on my local machine.

**Comment by alalazo** [comment]
@michaelkuhn Are you satisfied with what got in in #6312? Can we close this?

**Comment by michaelkuhn** [comment]
My approach is still a bit faster because I skip cleaning up the orphaned `.pyc` files and only import what is necessary. #6312 should do the trick, though. Thanks!

**Pull request closed without merging**